### ISSUE TITLE
False positive missing_const_for_thread_local

### ISSUE THREAD

**Issue by BenjaminBrienen** [rust-lang/rust-clippy#13422]
### Summary

missing_const_for_thread_local is already solved, but it triggers.

See https://github.com/bevyengine/bevy/issues/15241

### Lint Name

missing_const_for_thread_local

### Reproducer

I tried this code:

```rust
#[cfg(feature = "debug_stack")]
thread_local! {
    /// The thread-local [`TypeInfoStack`] used for debugging.
    ///
    /// [`TypeInfoStack`]: crate::type_info_stack::TypeInfoStack
    pub(super) static TYPE_INFO_STACK: std::cell::RefCell<crate::type_info_stack::TypeInfoStack> = const { std::cell::RefCell::new(
        crate::type_info_stack::TypeInfoStack::new()
    ) };
}
```

I saw this happen:

```
warning: initializer for `thread_local` value can be made `const`
  --> crates\bevy_reflect\src\serde\de\error_utils.rs:5:1
   |
5  | / thread_local! {
6  | |     /// The thread-local [`TypeInfoStack`] used for debugging.
7  | |     ///
8  | |     /// [`TypeInfoStack`]: crate::type_info_stack::TypeInfoStack
...  |
11 | |     ) };
12 | | }
   | |_^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_thread_local
   = note: `#[warn(clippy::missing_const_for_thread_local)]` on by default
   = note: this warning originates in the macro `$crate::thread::local_impl::thread_local_inner` which comes from the expansion of the macro `thread_local` (in Nightly builds, run with -Z macro-backtrace for more info)


```

I expected to see this happen:
No lint

### Version

rustc stable-x86_64-pc-windows-gnu

**Comment by BenjaminBrienen** [comment]
There are 3 examples of this false positive in the bevy repository. It doesn't trigger the false positive in the CI, which makes me think it's to do with the gnu vs msvc.

**Comment by BenjaminBrienen** [comment]
It does not trigger when I use msvc

**Comment by J-ZhengLi** [comment]
even the supposedly "good" example from this lint's documentation triggers it on windows-gnu target, this is very weird...

**Comment by rib** [comment]
I have some thread local variables declared like this that only trigger this false positive warning with `cargo clippy --target=aarch64-linux-android`

```rust
thread_local! {
    static THREAD_ATTACHMENT: Cell<*mut jni_sys::JNIEnv> = const { Cell::new(std::ptr::null_mut()) };
    static THREAD_GUARD_NEST_LEVEL: Cell<usize> = const { Cell::new(0) };
}
```

when running clippy for the host target I don't get this false positive warning

**Comment by jschwe** [comment]
I'm also getting a lot of such false positives in servo with clippy and `--target=aarch64-unknown-linux-ohos` which also don't appear when running clippy for host targets.

**Comment by xtqqczze** [comment]
Seeing this false positive for `aarch64-linux-android` and `x86_64-pc-solaris` targets:

```rust
thread_local! {
    static CELL: Cell<usize> = const { Cell::new(0) };
}
```

**Comment by xtqqczze** [comment]
Test case:

```rust
#[cfg_attr(
    any(
        target_os = "android",
        target_os = "haiku",
        target_os = "illumos",
        all(target_os = "linux", target_env = "ohos"),
        target_os = "solaris",
        all(target_os = "windows", target_env = "gnu", not(target_abi = "llvm"))),
    expect(
        clippy::missing_const_for_thread_local,
        reason = "https://github.com/rust-lang/rust-clippy/issues/13422"
    )
)]
thread_local! {
    static CELL: Cell<usize> = const { Cell::new(0) };
}
```
